analytics/analytics/job_processor/__init__.py

```diff
@@ -0,0 +1,66 @@
+import json
+from datetime import timedelta
+
+import gitlab
+import sentry_sdk
+from celery import shared_task
+from dateutil.parser import isoparse
+from django.conf import settings
+from gitlab.v4.objects import Project, ProjectJob
+
+from analytics import setup_gitlab_job_sentry_tags
+from analytics.job_processor.artifacts import annotate_job_with_artifacts_data
+from analytics.job_processor.build_timings import create_build_timings
+from analytics.job_processor.prometheus import (
+    JobPrometheusDataNotFound,
+    PrometheusClient,
+)
+from analytics.models import Job
+
+
+def create_job(gl: gitlab.Gitlab, project: Project, gljob: ProjectJob) -> Job:
+    # Create base fields on job that are independent of where it ran
+    job = Job(
+        job_id=gljob.get_id(),
+        project_id=project.get_id(),
+        name=gljob.name,
+        started_at=isoparse(gljob.started_at),
+        duration=timedelta(seconds=gljob.duration),
+        ref=gljob.ref,
+        tags=gljob.tag_list,
+        aws=True,  # Default until proven otherwise
+    )
+
+    # Prometheus data will either be found and the job annotated, or not, and set aws to False
+    try:
+        PrometheusClient(settings.SPACK_PROMETHEUS_ENDPOINT).annotate_job(job=job)
+    except JobPrometheusDataNotFound:
+        job.aws = False
+        annotate_job_with_artifacts_data(gljob=gljob, job=job)
+
+    # Handle some extra operations if aws job
+    if job.aws:
+        job.pod.save()
+
+        # Because jobs are many-to-one with nodes, we need to handle a race condition for node creation
+        job.save_or_set_node()
+
+    # Save and return new job
+    job.save()
+    return job
+
+
+@shared_task(name="process_job")
+def process_job(job_input_data_json: str):
+    # Read input data and extract params
+    job_input_data = json.loads(job_input_data_json)
+    setup_gitlab_job_sentry_tags(job_input_data)
+
+    # Retrieve project and job from gitlab API
+    gl = gitlab.Gitlab(settings.GITLAB_ENDPOINT, settings.GITLAB_TOKEN)
```

Comment:
This can be a future project, but we should look at setting `retry_transient_errors` and maybe using a global gitlab client (https://python-gitlab.readthedocs.io/en/stable/api-usage-advanced.html#transient-errors). We're already seeing how occasional failures creep in to the GitLab API with other jobs.

Reply:
I think that makes sense. What's the motivation for a global gitlab client? Is it just so it's initialized once when django is loaded, re-using the same client session in all parts of the application?

Reply:
Yeah, and it's just reusing the client in order to add convenience and reduce the chance that we're using differently-configured clients in different jobs.

Reply:
I addressed half of this in [7bb549b](https://github.com/spack/spack-infrastructure/pull/723/commits/7bb549b05eb8b1a49b877489d63d2bfd48ff88ae). The global gitlab client thing I would also like to implement at some point, but it's probably easier to do that in a follow-on PR, after all the concurrent work around the django app has been settled. I'll file an issue.